File: src_qt/rules/DataMiningNode.cpp
#include <QDebug>

#include "DataMiningNode.h"
#include "../Brain.h"


namespace CnotiMind
{

	DataMiningNode::DataMiningNode( const QString& event, const QString& value,
								   MemoryType memory, const QString& variable, const QString& position, Brain* brain, QObject* parent ) :
		RuleNode( brain, parent )//,
//		_event( event ),
//		_value( value ),
//		_operation( dataMiningOperation ),
//		_memory( memory ),
//		_variable( variable ),
//		_valueNumeric( value.toDouble( &_isValueNumeric ) ),
//		_position( position ),
//		_positionNumeric( position.toInt( &_isPositionNumeric ) )
	{

	}

	void DataMiningNode::exec()
	{
		if( isTrue( QHash<QString, QString>() ) )
		{
			if( _variable.isEmpty() )
			{
				// this should not happen, since the _variable should never be empty
				execChildren();
			}
			else
			{
				QHash<QString,QString> variables;
				variables.insert( _variable, _result.toString() );
				execChildren( variables );
			}
		}
	}

	void DataMiningNode::exec( QHash<QString, QString> &variables )
	{
		if( isTrue( variables ) )
		{
			if( !_variable.isEmpty() )
			{
				variables.insert( _variable, _result.toString() );
			}
			execChildren( variables );
		}
	}

	QString DataMiningNode::info( int depth ) const
	{
		QString info;

		info += space(depth) + "DataMining";

		return info + RuleNode::info(depth);
	}

	bool DataMiningNode::isTrue( const QHash<QString, QString> &variables )
	{
		bool valid;
		int position = 0;

		if( _isPositionNumeric )
		{
			position = _positionNumeric;
		}
		else // it should be a variable tag
		{
			QString positionStr = _position;
			bool ok;
			tagsToValue( positionStr, variables);

			// Test if it was possible to convert to number
			int p = positionStr.toInt( &ok );
			if( ok )
			{
				position = p;
			}
		}

		// If the values are numbers it should use the
		if( _isValueNumeric )
		{
			_result = _brain->dataMining( _operation, _event, _valueNumeric, position, _memory, &valid );

			return valid;
		}
		else // If the value is QString
		{
			_result = _brain->dataMining( _operation, _event, _value, position, _memory, &valid );

			return valid;
		}
	}

	DataMiningNode *DataMiningNode::fromXML(const QString &qName, const QXmlAttributes &atts, Brain *brain, QObject *parent)
	{
		qDebug() << "Inside";
		if( qName.compare( "Datamining", Qt::CaseInsensitive ) == 0 )
		{
			qDebug() << "Inside";

			QString event = atts.value( "event" );
			QString value = atts.value( "value" );
			qDebug() << "Inside";
			DataMiningOperation opDataMining = translateDataMiningOperator( atts.value( "operation" ) );
			qDebug() << "Inside";
			QString variable = atts.value( "variable" );
			QString position = atts.value( "position" );
			MemoryType memory = translateMemoryType( atts.value( "memory" ) );



			DataMiningNode* n = new DataMiningNode( event, value, memory,
													variable, position, brain, parent );
			qDebug() << "Data: " << n;
			return n;
		}
		return NULL;
	}
}

File: src_ios/CnotiMind/DataMiningNode.m
//
//  ConditionEmotionNode.m
//  CnotiMind
//
//  Created by Gonçalo Rodrigues on 11/05/09.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "DataMiningNode.h"

#import "Brain.h"
#import "Perception.h"

@implementation DataMiningNode


- (id) initWithEventAndValueAndOperatorAndMemoryAndVariableAndPositionAndBrainAndParent: (NSString*)aEvent 
                                                                                  value:(NSString*)aValue 
                                                                               operator:(enum DataMiningOperation)aDataMiningOperation 
                                                                                 memory:(enum MemoryType)aMemory
                                                                               variable:(NSString*)aVariable
                                                                               position:(NSString*)aPosition
                                                                                  brain:(Brain*)aBrain 
                                                                                 parent:(id)aParent
{
    
    if ( self == [super initWithBrainAndParent:aBrain parent:aParent] )
    {
        _event = (aEvent == nil) ? [[NSString alloc] initWithString:@""] : [[NSString alloc] initWithString:aEvent];
        _value = (aValue == nil) ? [[NSString alloc] initWithString:@""] : [[NSString alloc] initWithString:aValue];
        
        _operation = aDataMiningOperation;
        _memory = aMemory;
        _variable = [[NSString alloc] initWithString:aVariable];
        _valueNumeric = [NSNumber numberWithFloat:[aValue floatValue]];

        if (aPosition != nil) {
            _position = [[NSString alloc] initWithString:aPosition];
            
            _positionNumeric = [aPosition intValue];
            
            if ([self isNumeric:aPosition]) {
                _isPositionNumeric = TRUE;
            }
            else {
                _isPositionNumeric = FALSE;
            }            
        }
        else {
            _position = [[NSString alloc] initWithString:@""];
            _isPositionNumeric = FALSE;
//            _positionNumeric = [aPosition intValue];
        }
        
        
    }
    return self;
}


- (void) exec
{
    if( [self isTrue] )
    {
        if( [_variable length]==0 )
        {
            // this should not happen, since the _variable should never be empty
            [self execChildren];
        }
        else
        {
            NSMutableDictionary* variables = [[NSDictionary alloc] init];
            [variables setObject:_result forKey:_variable];
            
            //  TODO    
            [self execChildren: variables];
        }
    }
}


- (void) exec:(NSMutableDictionary*)aVariables
{
    if ([self isTrue: aVariables]) {
        
        if ( [_variable length] != 0 ) {
            [aVariables setObject:_result forKey:_variable];
        }
        
        [self execChildren];
    }
    
}


- (NSString*) info:(int)aDepth
{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ DataMining", space];
    info = [info stringByAppendingFormat:@"%@", [super info:aDepth]];
    
    return info;
}


- (BOOL) isTrue
{
    BOOL valid;
    
    // If the values are numbers it should use the
    if( _isValueNumeric )
    {
        _result = [_brain dataMining: _operation event:_event value:_valueNumeric memoryType:_memory valid:&valid];
        return valid;
    }
    else // If the value is QString
    {
        _result = [_brain dataMining: _operation event:_event value:_value memoryType:_memory valid:&valid];
        
        return valid;
    }
}

/*
 Check if in the value string there are variable or property tags
 
 If it is a variable tag try to get the value from the variables lists.
 
 Variables tags are betweeen square brackets [].
 Property tags are between curl brackets {}.
 
 If the variable/property tag is not found, it is replaced by an empty string.
 */

- (BOOL) isTrue:(NSMutableDictionary*)aVariables
{
    
    BOOL valid;
    int position = 0;
    
    if (_isPositionNumeric) {
        position = _positionNumeric;
    }
//    It sould be a variable tag
    else { 
        NSString* positionStr = [NSString stringWithFormat:@"%@", _position];
        BOOL ok;
        
        [self tagsToValue:&positionStr variables:&aVariables];
        
        int p = [positionStr intValue];
        position = p;
    }
        
    // If the values are numbers it should use the
    if( _isValueNumeric )
    {
        _result = [_brain dataMining: _operation event:_event value:_valueNumeric memoryType:_memory valid:&valid];
        return valid;
    }
    else // If the value is QString
    {
        _result = [_brain dataMining: _operation event:_event value:_value memoryType:_memory valid:&valid];
        
        return valid;
    }
}

- (void) dealloc
{
    [_event release];
    [_value release];
    [_variable release];
    
    [super dealloc];
}


@end
